### ISSUE TITLE
Provide a migration for Jenkins+GitLab to `Integrated code lifecycle`

### ISSUE THREAD

**Issue by reschandreas** [ls1intum/Artemis#8073]
### Is your feature request related to a problem?

As we're migrating away from Atlassian to `ICL`. I would like to list the steps that would be required, from my point of view, for a similar migration from Jenkins. I'll list which hurdles, with an estimation, we need to tackle before we can offer a similar migration as in #7842 and #7855

1. The Jenkins pipeline may need to be loaded from both Jenkins and Artemis, as Aeolus currently cannot handle it. If we assume that only the build plan in Artemis is our source of truth, the Aeolus API needs modification for parsing. This modification does not require changing the current API and can be done in approximately one day.
2. Parsing the pipelines is the most time-consuming task. The current templates may take 3-4 days to build a rudimentary parser for basic cases. However, handling edge cases and supporting old tasks with different formats could significantly increase the effort, roughly estimated at one to two weeks. This needs to be done in [aeolus](https://github.com/ls1intum/Aeolus).
3. Depending on the usage of Blackbox Maven, complexity increases. At least this type of exercise requires script rewriting due to pipeline method calls with various options. This involves distinguishing between student and solution build plans, which is not currently possible in LocalCI. Adapting to these complexities, including Docker parameter considerations and project-type-specific adjustments, may add several days and require extensive debugging.
Additionally, parsing the results of these tasks in LocalCI (presumably JSON files) is missing; I'd estimate that this will take at least one week. I added a platform-independent template in #8029, but some restrictions still apply, such as the parameters passed to the docker container in the student build plans.

### Describe the solution you'd like

What do you think of the steps and estimations?

### Describe alternatives you've considered

Given that this is an extensive new feature that requires a lot of hard work within [aeolus](https://github.com/ls1intum/Aeolus) and Artemis, I'd like to bring another idea to the table: 
Make Aeolus mandatory for Jenkins + GitLab from a certain version onwards, storing the build plans in a platform-independent format (the windfile). This would allow for a very straightforward migration of new exercises, but at a later time when most exercises don't need any manual activities to be used with LocalCI.

But this would also limit the the working edit build plan feature already part of Artemis and used by other universities.

**Comment by b-fein** [comment]
> However, handling edge cases and supporting old tasks with different formats could significantly increase the effort, roughly estimated at one to two weeks.

I do not think you will be able to parse and translate existing `pipeline.groovy` files that deviate from the template into the Aeolus format:

The pipeline in this case can be arbitrary Groovy-Code with the addition of the Jenkins-specific-DSL/methods. This allows for the conditional or repeated  inclusion of more pipeline steps (`stage`s in the Jenkins terminology).

This is for example used to differentiate between the solution and student build plans (Maven Blackbox Template), but also allows for further customisation, e.g. for A/B grouping of students based on the hash of the repo name (idea proposed in #2812, maybe in use in the CodeAbility project).

---

> 3.

This is not limited to the use of the Blackbox Maven template. For us, we have many exercises where the base-Maven template was used to set up the initial project, but the pipeline has been completely customised afterwards. For example, in a software testing course, a pipeline might look like

<details>
<summary>Pipeline Groovy Example</summary>

```groovy
dockerImage = "ghcr.io/uni-passau-artemis/artemis-java-firefox:21"
dockerFlags = "..."

void testRunner() {
    setDockerFlags()

    docker.image(dockerImage).inside(dockerFlags) { c ->
        runTestSteps()
    }
}

private void setDockerFlags() {
    // … similar to Maven Blackbox
}

private void runTestSteps() {
    catchError {
        timeout(time: 15, unit: 'MINUTES') {
           studentTests()
        }
    }

    catchError {
        timeout(time: 15, unit: 'MINUTES') {
            functionalTests()
        }
    }

    catchError {
        pythonChecks()
    }
}

private void studentTests() {
    stage('Execute Cucumber Tests Written by Student') {
        // We need to set HOME for this command, because otherwise it is bound to `/`, which is not
        // writable for the user that runs the job inside the Docker container.
        sh "HOME=/tmp mvn --no-transfer-progress -Pintegration-tests clean verify || true"
        sh "cat target/failsafe-reports/*.txt > cucumber.log"
        sh 'mkdir -p results'
        sh 'cp target/failsafe-reports/failsafe-summary.xml $WORKSPACE/results/cucumber.xml || true'
    }
}

private void functionalTests() {
    stage('Execute Function Tests by Us') {
        // We need to set HOME for this command, because otherwise it is bound to `/`, which is not
        // writable for the user that runs the job inside the Docker container.
        sh "HOME=/tmp mvn --no-transfer-progress -Psystem-tests clean verify || true"
        sh "cat target/failsafe-reports/*.txt > functional_tests.log"
    }
}

private void pythonChecks() {
    stage('Execute BDD Process Check') {
        sh "python3 tdd_process_checker.py ./assignment/"
    }
    stage("Export Cucumber Results") {
        sh "python3 cucumber_results.py"
    }
}

void postBuildTasks() {
    // …
}

return this
```

</details>

---

Some concrete examples we use the pipelines for that are not supported by the local CI workflows at all as far as I know:

- dynamically set up the flags a container should be started with, i.e. run some setup code outside the container
- start sidecar containers for a database (similar to https://www.jenkins.io/doc/book/pipeline/docker/#running-sidecar-containers)
- run different stages in different containers

**Comment by b-fein** [comment]
Migration of GitLab to LocalVCS exists with #8972.
